core/src/main/scala/org/bitcoins/core/p2p/NetworkPayload.scala

```diff
@@ -166,13 +166,27 @@ object GetDataMessage extends Factory[GetDataMessage] {
     GetDataMessage(Seq(inventory))
 }
 
+sealed trait ExpectsResponse {
```

Comment:
So I needed a way to if the peer failed to respond to a query for eg `getheaders` etc that is used in sync, so that I can switch to a different peer and continue sync from there. For `NetworkPayload` that extends this, an `ExpectReponse` message is sent to the `P2PClient` right after sending the message. The message initiates a timer in `PeerMessageReceiver` that notifies if the message timed out.

Reply:
Was this common when testing? This is an interesting blog post about the bitcoin p2p network possibly slowing down: https://blog.lopp.net/is-bitcoin-network-slowing-down/

it details the logic bitcoin core has about detecting peers that are stalling, probably worth a read. It seems the behavior detailed there is what you experienced as well?